main.go

```diff
@@ -0,0 +1,205 @@
+	cfg, err := config.NewFromCFHome()
+	if err != nil {
+		return &client.Client{}, err
+	}
+
+	cfc, err := client.New(cfg)
+	if err != nil {
+		return &client.Client{}, err
+	}
+
+	return cfc, nil
+}
+
+func retrieveDomains(cfc *client.Client, domainName string) ([]*resource.Domain, error) {
+	domainOpts := client.NewDomainListOptions()
+	domainOpts.Names.Values = append(domainOpts.Names.Values, domainName)
+	domains, err := cfc.Domains.ListAll(context.Background(), domainOpts)
+	if err != nil {
+		return nil, err
+	}
+	return domains, nil
+}
+
+func findDomain(cfc *client.Client, query string) (*resource.Domain, string, *url.URL, error) {
+	routeUrl, err := url.Parse(query)
+	if err != nil {
+		return &resource.Domain{}, "", &url.URL{}, err
+	}
+	if routeUrl.Scheme == "" {
+		return &resource.Domain{}, "", routeUrl, fmt.Errorf("please provide the url including the scheme")
+	}
+
+	domains, err := retrieveDomains(cfc, routeUrl.Hostname())
+	if err != nil {
+		return &resource.Domain{}, routeUrl.Hostname(), routeUrl, err
+	}
+
+	if len(domains) == 0 {
+		urlParts := strings.SplitN(routeUrl.Hostname(), ".", 2)
+		if len(urlParts) < 2 {
+			return &resource.Domain{}, "", routeUrl, fmt.Errorf("'%s' is not a domain", routeUrl.Hostname())
+		}
+		hostName := urlParts[0] //Subdomain is not empty
+		domainName := urlParts[1]
+		domains, err = retrieveDomains(cfc, domainName)
+		if len(domains) == 0 {
+			return &resource.Domain{}, hostName, routeUrl, fmt.Errorf("error retrieving apps: route not found, domain '%s' is unknown", domainName)
+		}
+		return domains[0], hostName, routeUrl, nil
+	}
+	return domains[0], routeUrl.Hostname(), routeUrl, nil
+}
+
+func findRoute(cfc *client.Client, query string) (*resource.Route, error) {
+	domain, hostName, routeUrl, err := findDomain(cfc, query)
+
+	opts := client.NewRouteListOptions()
+	opts.Hosts.Values = append(opts.Hosts.Values, hostName)
+	opts.DomainGUIDs.Values = append(opts.DomainGUIDs.Values, domain.GUID)
+	opts.Paths.Values = append(opts.Paths.Values, routeUrl.Path)
+
+	routes, err := cfc.Routes.ListAll(context.Background(), opts)
+	if err != nil {
+		return &resource.Route{}, err
+	}
+
+	if len(routes) == 0 { // Wildcard search
+		opts.Hosts.Values = append(opts.Hosts.Values, "*")
+		routes, err = cfc.Routes.ListAll(context.Background(), opts)
+		if err != nil {
+			return &resource.Route{}, err
+		}
+		if len(routes) == 0 {
+			return &resource.Route{}, fmt.Errorf("error retrieving apps: route '%s' not found", routeUrl.Hostname())
+		}
+	}
+
+	return routes[0], nil
+}
+
+func lookup(cfc *client.Client, route *resource.Route, target bool, cliConnection plugin.CliConnection) error {
+	if route.Destinations == nil || len(route.Destinations) == 0 {
+		return fmt.Errorf("route not bound to any applications")
+	}
+
+	destinationCount := 0
+	for _, destination := range route.Destinations {
+		app, err := cfc.Applications.Get(context.Background(), *destination.App.GUID)
```

Comment:
Just as note: the old implementation just returned the first result. This repeats the org/space for each app bound.

Looping over all destinations, and querying each app individually, might add to the runtime and stress on the Cloud Controller API.

Unless we critically need to list all apps, or if we can get the apps for multiple GUIDs, we should consider some "short circuit" mode where only the first result is queried and shown.

Reply:
Done